File: src/clu/sources/clu.py
import datetime
import getpass
import logging
import os
import sys

from clu import __about__
from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts, Tier
from clu.sources import Source

log = logging.getLogger(__name__)


def get_rfc3339_timestamp() -> str:
    """Get the current time in RFC 3339 format."""
    now_utc = datetime.datetime.now(datetime.timezone.utc)
    return now_utc.isoformat(sep="T", timespec="seconds")


# This silly function makes testing easier - its easier to patch() out
def raw_clu_metadata() -> dict:
    """Return info about CLU itself (mostly runtime)."""

    return {
        "sys.argv": sys.argv,
        "about.version": __about__.__version__,
        "sys.executable": sys.executable,
        "sys.version_info": sys.version_info,
        "os.getcwd": os.getcwd(),
        "getpass.getuser": getpass.getuser(),
        "datestamp": get_rfc3339_timestamp(),
    }


class Clu(Source):
    def provides(self, provides: Provides) -> None:
        """Define the provider map for CLU."""
        for key in [
            "clu.binary",
            "clu.version",
            "clu.python.binary",
            "clu.python.version",
            "clu.cmdline",
            "clu.cwd",
            "clu.user",
            "clu.date",
        ]:
            provides[key] = self

    def requires(self, requires: Requires) -> None:
        # No specific requirements for clu group
        pass

    def parse(self, facts: Facts) -> None:
        """Return info about CLU itself (mostly runtime)."""

        metadata = raw_clu_metadata()
        facts.add(Tier.TWO, "clu.binary", metadata["sys.argv"][0])
        facts.add(Tier.ONE, "clu.version", metadata["about.version"])
        facts.add(Tier.TWO, "clu.python.binary", metadata["sys.executable"])
        facts.add(
            Tier.ONE, "clu.python.version", ".".join(map(str, metadata["sys.version_info"][:3]))
        )
        facts.add(Tier.TWO, "clu.cmdline", " ".join(metadata["sys.argv"]))
        facts.add(Tier.THREE, "clu.cwd", metadata["os.getcwd"])
        facts.add(Tier.THREE, "clu.user", metadata["getpass.getuser"])
        facts.add(Tier.TWO, "clu.date", metadata["datestamp"])

File: pkg/sources/clu.go
package sources

import (
	"os"
	"os/user"
	"runtime"
	"strings"
	"time"

	pkg "github.com/huntermatthews/clu/pkg"
)

// Clu provides runtime-related facts about the running CLI and environment.
type Clu struct{}

func (c *Clu) Provides(p pkg.Provides) {
	for _, k := range []string{
		"clu.binary",
		"clu.version",
		"clu.python.binary",
		"clu.python.version",
		"clu.cmdline",
		"clu.cwd",
		"clu.user",
		"clu.date",
	} {
		p[k] = c
	}
}

func (c *Clu) Requires(r *pkg.Requires) { /* no external requirements */ }

func (c *Clu) Parse(f *pkg.Facts) {
	// Binary name (argv0)
	argv0 := os.Args[0]
	f.Add(pkg.TierTwo, "clu.binary", argv0)

	// Version from about.go
	f.Add(pkg.TierOne, "clu.version", pkg.Version)

	// Python analogs: in Go, provide placeholders: binary is the current executable path
	// and version is runtime.Version()
	f.Add(pkg.TierTwo, "clu.python.binary", argv0)
	f.Add(pkg.TierOne, "clu.python.version", runtime.Version())

	// Command line
	f.Add(pkg.TierTwo, "clu.cmdline", strings.Join(os.Args, " "))

	// Working directory
	cwd, _ := os.Getwd()
	f.Add(pkg.TierThree, "clu.cwd", cwd)

	// User
	u, _ := user.Current()
	if u != nil && u.Username != "" {
		f.Add(pkg.TierThree, "clu.user", u.Username)
	} else {
		f.Add(pkg.TierThree, "clu.user", "")
	}

	// RFC3339 timestamp (UTC)
	f.Add(pkg.TierTwo, "clu.date", time.Now().UTC().Format(time.RFC3339))
}
